service/storage_service.cc

```diff
@@ -5398,39 +5426,48 @@ void storage_service::init_messaging_service(sharded<service::storage_proxy>& pr
                 auto partition_range = dht::partition_range::make_singular(key);
                 auto rs = co_await db::system_keyspace::query_mutations(
                     db, db::system_keyspace::NAME, db::system_keyspace::CDC_GENERATIONS_V3, partition_range);
                 if (rs->partitions().size() != 1) {
                     on_internal_error(slogger, ::format(
                         "pull_raft_topology_snapshot: expected a single partition in CDC generation query,"
-                        ", got {} (generation ID: {})", rs->partitions().size(), *curr_cdc_gen_id));
+                        ", got {} (generation UUID: {})", rs->partitions().size(), uuid));
                 }
-                cdc_generation_mutation.emplace(rs->partitions().begin()->mut().unfreeze(s));
-            }
+                co_return rs->partitions().begin()->mut().unfreeze(s);
+            }, std::ref(ss), std::ref(db));
+
+            // We don't need to fetch this data under group0 apply mutex, it's immutable.
+            // We only need to include the current CDC generation data and the next one if it is being broadcasted,
+            // because nodes only load whatever `curr_cdc_gen_id` and `in_progress_cdc_gen_id` point to in topology.
+            //
+            // FIXME: when we bootstrap nodes in quick succession, the timestamp of the newest CDC generation
+            // may be for some time larger than the clocks of our nodes. The last bootstrapped node
+            // will only receive the newest CDC generation and not earlier ones, so it will only be able
+            // to coordinate writes to CDC-enabled tables after its clock advances to reach the newest
+            // generation's timestamp. In other words, it may not be able to coordinate writes for some
+            // time after bootstrapping and drivers connecting to it will receive errors.
+            // To fix that, we could store in topology a small history of recent CDC generation IDs
+            // (garbage-collected with time) instead of just the last one, and load all of them.
+            // Alternatively, a node would wait for some time before switching to normal state.
+            std::optional<canonical_mutation> cdc_generation_mutation = curr_cdc_gen_id
+                ? std::optional<canonical_mutation>{co_await load_mutations(curr_cdc_gen_id->id)}
+                : std::nullopt;
+
+            std::optional<canonical_mutation> cdc_in_progress_generation_mutation = in_progress_cdc_gen_id
+                ? std::optional<canonical_mutation>{co_await load_mutations(*in_progress_cdc_gen_id)}
+                : std::nullopt;
 
             co_return raft_topology_snapshot{
                 .topology_mutations = std::move(topology_mutations),
                 .cdc_generation_mutation = std::move(cdc_generation_mutation),
+                .cdc_in_progress_generation_mutation = std::move(cdc_in_progress_generation_mutation),
```

Comment:
Why do you need a separate field for it?

Reply:
This part of the code was removed due to a change of topology snapshots. Closing conversation.